working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,379 @@
+
+  /// Same as [super.register] but allows to specify an [externalSize] to
+  /// guide GC heuristics.
+  void register(F value,
+                Pointer finalizationToken,
+                {Object? unregisterToken, int externalSize});
+}
+```
+
+`FinalizationRegistry` was directly modeled after its
+[JavaScript counterpart][MDN FinalizationRegistry] and only supports
+asynchronous finalization, while `NativeFinalizationRegistry` is added to
+`dart:ffi` to allow eager synchronous finalization.
+
+Note differences in API between `FinalizationRegistry` and
+`NativeFinalizationRegistry`:
+
+- `NativeFinalizationRegistry` requires objects which are registered with it
+to implement `Finalizable` interface, which serves as a marker instructing
+optimizing compiler to provide stronger liveness guarantees for an object. This
+interface is our solution to the problem of _premature finalization_.
+- `NativeFinalizationRegistry` is constructed with a _native function_ as a
+callback rather than a Dart function. This is done to guarantee that eager
+synchronous execution of a finalization callback is not going to produce any
+side-effects observable from the pure Dart code.
+
+### Isolate-independent native functions
+
+`dart:ffi` allows developers to construct a native function pointer to a
+static function defined in Dart via [`Pointer.fromFunction`] constructor.
+This constructor essentially returns a pointer to a _trampoline_ which
+follows C ABI and can be called from native code. When invoked such trampoline
+will perform transition from native code into Dart code, marshal arguments,
+call the target static function, marshaling the result it returns and then
+transition back into native code. There is an implicit expectation baked into
+this process: calling a native-to-Dart trampoline will only succeed if
+there is a Dart isolate associated with the thread on which we perform the call
+and this isolate is in the state which allows reentrancy.
+
+Such isolate-dependent function can't be used as a finalization callback because
+finalization callbacks should be callable in contexts when there is no current
+isolate at all or isolates are not allowing entering into Dart code.
+
+It seems thus reasonable to restrict `NativeFinalizationRegistry` constructor
+in a way that would reject function pointers which are pointing to
+native-to-Dart FFI trampolines.
+
+This restriction however means that users might be required to write native code
+to implement their finalizers. Consider for example [`mmap`/`munmap`][mmap API]
+POSIX APIs.
+
+```cpp
+// mmap -- allocate memory, or map files or devices into memory
+void *mmap(void *addr, size_t len, int prot, int flags, int fd, off_t offset);
+
+// munmap -- remove a mapping
+int munmap(void *addr, size_t len);
+```
+
+A Dart developer will be able to call `mmap` and `munmap` through FFI, but they
+will not be able to use `munmap` directly as a finalizer, because it does not
```

Comment:
If we take the approach of 

```dart
abstract class NativeFinalizationRegistry<F extends Finalizable, C extends Function>
    extends FinalizationRegistry<F, Pointer> {
  external factory NativeFinalizationRegistry(Pointer<NativeFunction<C>> callback,
      int tokenArgumentIndex, List otherArguments);
}
```

the developer would be able to pass in the size to the constructor.

(Of course that is less than ideal, we'd want that to be an argument to `register` instead.)

Reply:
FWIW I think "Isolate-independent native functions" can be built in stages: we don't have to support a lot of things to make it usable, e.g. we could build something that compiles Dart version of the `FinalizeMapping` in an isolate independent way rather simply and that would be enough to support `package:win32` and other similarly trivial use cases.

The reason why I keep this out of this proposal is not because I think that we should not implement it (in fact I think we should implement it and ship it as part of the finalizers - it is as you yourself highlight a prerequisite to make finalizers more usable for FFI specific usecases), but rather to keep decisions of how we define and ship it outside of scope for this proposal, which will consequently remain purely focused on finalizers.

Reply:
I am going to sketch the implementation of the isolate independent functions to confirm that we could get it implemented in a reasonable time.